repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneRequestHandler.scala

```diff
@@ -0,0 +1,314 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.filters.keystonev2
+
+import java.io.InputStream
+import java.util.{Calendar, GregorianCalendar}
+import javax.servlet.http.HttpServletResponse._
+import javax.ws.rs.core.MediaType
+
+import com.fasterxml.jackson.core.JsonProcessingException
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import org.apache.http.HttpHeaders
+import org.apache.http.client.utils.{DateUtils, URIBuilder}
+import org.joda.time.format.ISODateTimeFormat
+import org.openrepose.commons.utils.http.{CommonHttpHeader, ServiceClientResponse}
+import org.openrepose.core.services.serviceclient.akka.AkkaServiceClient
+import play.api.libs.functional.syntax._
+import play.api.libs.json.Reads._
+import play.api.libs.json._
+
+import scala.collection.JavaConverters._
+import scala.io.Source
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Contains the functions which interact with the Keystone API.
+ */
+class KeystoneRequestHandler(identityServiceUri: String, akkaServiceClient: AkkaServiceClient, traceId: Option[String])
+  extends LazyLogging {
+
+  import KeystoneRequestHandler._
+
+  /**
+   * Call to the Keystone service to get the admin token
+   * @return a Successful token, or a Failure
+   */
+  final def getAdminToken(adminUsername: String, adminPassword: String): Try[String] = {
+    //authenticate, or get the admin token
+    val authenticationPayload = Json.obj(
+      "auth" -> Json.obj(
+        "passwordCredentials" -> Json.obj(
+          "username" -> adminUsername,
+          "password" -> adminPassword
+        )
+      )
+    )
+
+    val tokenEndpointUri = new URIBuilder().setHost(identityServiceUri).setPath(TOKEN_ENDPOINT).build().toString()
```

Comment:
Side note: the thing I like about the JAX-RS UriBuilder is that you can append to the path. So you can do something like:
new UriBuilder()
.path("/foo")
.path("/bar")
.build()

Whereas we can only set the path with this UriBuilder. So we still have to manage the '/'s in the path ourselves.